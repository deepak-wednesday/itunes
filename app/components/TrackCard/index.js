/**
 *
 * TrackCard
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card } from 'antd';
import { PlayCircleTwoTone, RightCircleTwoTone, PauseCircleTwoTone } from '@ant-design/icons';
import { Link } from 'react-router-dom';
import * as colors from '@app/themes/colors';
import If from '@components/If';

const { Meta } = Card;

const CustomCard = styled(Card)`
  && {
    width: 18em;
    margin: 1rem 0;
    height: ${(props) => (props.height ? props.height : 30)}em;
    background-color: ${colors.trackCardColor};
  }
`;
const ImageCard = styled(Card)`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 15em;
  height: ${(props) => (props.height ? props.height : 24)}em;
`;
const TextCard = styled(Meta)`
  && {
    margin: 0.3rem !important;
    width: 15em;
    padding: 1em;
  }
`;
const Image = styled.img`
  width: 15em;
  height: ${(props) => (props.height ? props.height : 18)}em;
`;
const CustomPause = styled(PauseCircleTwoTone)`
  font-size: 2rem;
  padding: 1rem;
  cursor: pointer;
`;
const CustomPlay = styled(PlayCircleTwoTone)`
  font-size: 2rem;
  padding: 1rem;
  cursor: pointer;
`;
const CustomRight = styled(RightCircleTwoTone)`
  font-size: 2rem;
  padding: 1rem;
  cursor: pointer;
`;
const CustomButton = styled.div`
  dispaly: flex;
  justify-content: center;
  align-items: center;
`;

export function TrackCard({ item, currentPlayingId, isPlaying, onPlay, onPause }) {
  const { trackName, artistName, artworkUrl100, previewUrl, trackId } = item;
  return (
    <CustomCard data-testid="track-card">
      <Link to={`/track/${trackId}`}>
        <ImageCard>
          <Image alt="artwork" src={artworkUrl100} />
          <TextCard data-testid="text-card" title={trackName} description={artistName} />
        </ImageCard>
      </Link>
      <CustomButton>
        <If condition={isPlaying && item.trackId === currentPlayingId}>
          <CustomPause data-testid="pause-button" onClick={() => onPause(previewUrl)} />
        </If>
        <If condition={!(isPlaying && item.trackId === currentPlayingId)}>
          <CustomPlay data-testid="play-button" onClick={() => onPlay(previewUrl, trackId)} />
        </If>
        <CustomRight />
      </CustomButton>
    </CustomCard>
  );
}

TrackCard.propTypes = {
  height: PropTypes.number,
  item: PropTypes.object,
  onPause: PropTypes.func,
  onPlay: PropTypes.func,
  currentPlayingId: PropTypes.number,
  isPlaying: PropTypes.bool,
  trackId: PropTypes.number
};

export default TrackCard;
