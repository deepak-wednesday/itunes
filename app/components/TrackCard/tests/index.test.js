/**
 *
 * Tests for TrackCard
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderWithIntl, timeout } from '@utils/testUtils';
import TrackCard from '../index';
import testData from '../testData';

describe('<TrackCard />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TrackCard item={testData} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackCard component', () => {
    const { getAllByTestId } = renderWithIntl(<TrackCard item={testData} />);
    expect(getAllByTestId('track-card').length).toBe(1);
  });

  it('should work for play accordingly', async () => {
    const { getByTestId } = renderWithIntl(<TrackCard item={testData} onPlay={jest.fn()} />);
    fireEvent.click(getByTestId('play-button'));
    await timeout(500);
    expect(getByTestId('play-button')).toBeEnabled();
  });
  it('should render trackname & description', () => {
    const { getByTestId } = renderWithIntl(<TrackCard item={testData} />);
    expect(getByTestId('text-card')).toHaveTextContent(testData.trackName);

    expect(getByTestId('text-card')).toHaveTextContent(testData.artistName);
  });
});
