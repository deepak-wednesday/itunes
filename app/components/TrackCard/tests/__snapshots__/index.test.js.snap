// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered TrackCard__CustomCard-sc-10bk196-0 hZPEMI"
      data-testid="track-card"
    >
      <div
        class="ant-card-body"
      >
        <a
          href="/track/1469577741"
        >
          <div
            class="ant-card ant-card-bordered TrackCard__ImageCard-sc-10bk196-1 jvaOyH"
          >
            <div
              class="ant-card-body"
            >
              <img
                alt="artwork"
                class="TrackCard__Image-sc-10bk196-3 ioNbZe"
                src="https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/ae/4c/d4/ae4cd42a-80a9-d950-16f5-36f01a9e1881/source/100x100bb.jpg"
              />
              <div
                class="ant-card-meta TrackCard__TextCard-sc-10bk196-2 cIXUut"
                data-testid="text-card"
              >
                <div
                  class="ant-card-meta-detail"
                >
                  <div
                    class="ant-card-meta-title"
                  >
                    Upside Down
                  </div>
                  <div
                    class="ant-card-meta-description"
                  >
                    Jack Johnson
                  </div>
                </div>
              </div>
            </div>
          </div>
        </a>
        <div
          class="TrackCard__CustomButton-sc-10bk196-7 cOpOKr"
        >
          <span
            aria-label="play-circle"
            class="anticon anticon-play-circle TrackCard__CustomPlay-sc-10bk196-5 jxbcWi"
            data-testid="play-button"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="play-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                fill="#1890ff"
              />
              <path
                d="M512 140c-205.4 0-372 166.6-372 372s166.6 372 372 372 372-166.6 372-372-166.6-372-372-372zm164.1 378.2L457.7 677.1a8.02 8.02 0 01-12.7-6.5V353a8 8 0 0112.7-6.5l218.4 158.8a7.9 7.9 0 010 12.9z"
                fill="#e6f7ff"
              />
              <path
                d="M676.1 505.3L457.7 346.5A8 8 0 00445 353v317.6a8.02 8.02 0 0012.7 6.5l218.4-158.9a7.9 7.9 0 000-12.9z"
                fill="#1890ff"
              />
            </svg>
          </span>
          <span
            aria-label="right-circle"
            class="anticon anticon-right-circle TrackCard__CustomRight-sc-10bk196-6 iTvnfB"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 140c-205.4 0-372 166.6-372 372s166.6 372 372 372 372-166.6 372-372-166.6-372-372-372zm154.7 378.4l-246 178c-5.3 3.8-12.7 0-12.7-6.5V643c0-10.2 4.9-19.9 13.2-25.9L566.6 512 421.2 406.8c-8.3-6-13.2-15.6-13.2-25.9V334c0-6.5 7.4-10.3 12.7-6.5l246 178c4.4 3.2 4.4 9.7 0 12.9z"
                fill="#e6f7ff"
              />
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                fill="#1890ff"
              />
              <path
                d="M666.7 505.5l-246-178c-5.3-3.8-12.7 0-12.7 6.5v46.9c0 10.3 4.9 19.9 13.2 25.9L566.6 512 421.2 617.1c-8.3 6-13.2 15.7-13.2 25.9v46.9c0 6.5 7.4 10.3 12.7 6.5l246-178c4.4-3.2 4.4-9.7 0-12.9z"
                fill="#1890ff"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
